README.md
# MULTISALARYINPUT FOR [DOLIBARR ERP CRM](https://www.dolibarr.org)

## Features

Description of the module...

<!--
![Screenshot multisalaryinput](img/screenshot_multisalaryinput.png?raw=true "MultiSalaryInput"){imgmd}
-->

Other external modules are available on [Dolistore.com](https://www.dolistore.com).

## Translations

Translations can be completed manually by editing files into directories *langs*.

<!--
This module contains also a sample configuration for Transifex, under the hidden directory [.tx](.tx), so it is possible to manage translation using this service.

For more informations, see the [translator's documentation](https://wiki.dolibarr.org/index.php/Translator_documentation).

There is a [Transifex project](https://transifex.com/projects/p/dolibarr-module-template) for this module.
-->

<!--

## Installation

### From the ZIP file and GUI interface

- If you get the module in a zip file (like when downloading it from the market place [Dolistore](https://www.dolistore.com)), go into
menu ```Home - Setup - Modules - Deploy external module``` and upload the zip file.

Note: If this screen tell you there is no custom directory, check your setup is correct:

- In your Dolibarr installation directory, edit the ```htdocs/conf/conf.php``` file and check that following lines are not commented:

    ```php
    //$dolibarr_main_url_root_alt ...
    //$dolibarr_main_document_root_alt ...
    ```

- Uncomment them if necessary (delete the leading ```//```) and assign a sensible value according to your Dolibarr installation

    For example :

    - UNIX:
        ```php
        $dolibarr_main_url_root_alt = '/custom';
        $dolibarr_main_document_root_alt = '/var/www/Dolibarr/htdocs/custom';
        ```

    - Windows:
        ```php
        $dolibarr_main_url_root_alt = '/custom';
        $dolibarr_main_document_root_alt = 'C:/My Web Sites/Dolibarr/htdocs/custom';
        ```

### From a GIT repository

- Clone the repository in ```$dolibarr_main_document_root_alt/multisalaryinput```

```sh
cd ....../custom
git clone git@github.com:gitlogin/multisalaryinput.git multisalaryinput
```

### <a name="final_steps"></a>Final steps

From your browser:

  - Log into Dolibarr as a super-administrator
  - Go to "Setup" -> "Modules"
  - You should now be able to find and enable the module

-->

## Licenses

### Main code

GPLv3 or (at your option) any later version. See file COPYING for more information.

### Documentation

All texts and readmes are licensed under GFDL.


ChangeLog.md
# CHANGELOG MULTISALARYINPUT FOR [DOLIBARR ERP CRM](https://www.dolibarr.org)

## 1.0

Initial version


test/phpunit/MultiSalaryInputFunctionalTest.php
<?php
/* Copyright (C) 2007-2017 Laurent Destailleur  <eldy@users.sourceforge.net>
 * Copyright (C) 2023 SuperAdmin
 *
 * This program is free software: you can redistribute it and/or modify
 * it under the terms of the GNU General Public License as published by
 * the Free Software Foundation, either version 3 of the License, or
 * (at your option) any later version.
 *
 * This program is distributed in the hope that it will be useful,
 * but WITHOUT ANY WARRANTY; without even the implied warranty of
 * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
 * GNU General Public License for more details.
 *
 * You should have received a copy of the GNU General Public License
 * along with this program.  If not, see <https://www.gnu.org/licenses/>.
 */

/**
 * \file    test/phpunit/MultiSalaryInputFunctionalTest.php
 * \ingroup multisalaryinput
 * \brief   Example Selenium test.
 *
 * Put detailed description here.
 */

namespace test\functional;

use PHPUnit_Extensions_Selenium2TestCase_WebDriverException;

/**
 * Class MultiSalaryInputFunctionalTest
 *
 * Requires chromedriver for Google Chrome
 * Requires geckodriver for Mozilla Firefox
 *
 * @fixme Firefox (Geckodriver/Marionette) support
 * @todo Opera linux support
 * @todo Windows support (IE, Google Chrome, Mozilla Firefox, Safari)
 * @todo OSX support (Safari, Google Chrome, Mozilla Firefox)
 *
 * @package Testmultisalaryinput
 */
class MultiSalaryInputFunctionalTest extends \PHPUnit_Extensions_Selenium2TestCase
{
	// TODO: move to a global configuration file?
	/** @var string Base URL of the webserver under test */
	protected static $base_url = 'http://dev.zenfusion.fr';
	/**
	 * @var string Dolibarr admin username
	 * @see authenticate
	 */
	protected static $dol_admin_user = 'admin';
	/**
	 * @var string Dolibarr admin password
	 * @see authenticate
	 */
	protected static $dol_admin_pass = 'admin';
	/** @var int Dolibarr module ID */
	private static $module_id = 500000; // TODO: autodetect?

	/** @var array Browsers to test with */
	public static $browsers = array(
		array(
			'browser' => 'Google Chrome on Linux',
			'browserName' => 'chrome',
			'sessionStrategy' => 'shared',
			'desiredCapabilities' => array()
		),
		// Geckodriver does not keep the session at the moment?!
		// XPath selectors also don't seem to work
		//array(
		//    'browser' => 'Mozilla Firefox on Linux',
		//    'browserName' => 'firefox',
		//    'sessionStrategy' => 'shared',
		//    'desiredCapabilities' => array(
		//        'marionette' => true,
		//    ),
		//)
	);

	/**
	 * Helper function to select links by href
	 *
	 * @param  string  $value      Href
	 * @return mixed               Helper string
	 */
	protected function byHref($value)
	{
		$anchor = null;
		$anchors = $this->elements($this->using('tag name')->value('a'));
		foreach ($anchors as $anchor) {
			if (strstr($anchor->attribute('href'), $value)) {
				break;
			}
		}
		return $anchor;
	}

	/**
	 * Global test setup
	 * @return void
	 */
	public static function setUpBeforeClass()
	{
	}

	/**
	 * Unit test setup
	 * @return void
	 */
	public function setUp()
	{
		$this->setSeleniumServerRequestsTimeout(3600);
		$this->setBrowserUrl(self::$base_url);
	}

	/**
	 * Verify pre conditions
	 * @return void
	 */
	protected function assertPreConditions()
	{
	}

	/**
	 * Handle Dolibarr authentication
	 * @return void
	 */
	private function authenticate()
	{
		try {
			if ($this->byId('login')) {
				$login = $this->byId('username');
				$login->clear();
				$login->value('admin');
				$password = $this->byId('password');
				$password->clear();
				$password->value('admin');
				$this->byId('login')->submit();
			}
		} catch (PHPUnit_Extensions_Selenium2TestCase_WebDriverException $e) {
			// Login does not exist. Assume we are already authenticated
		}
	}

	/**
	 * Test enabling developer mode
	 * @return bool
	 */
	public function testEnableDeveloperMode()
	{
		$this->url('/admin/const.php');
		$this->authenticate();
		$main_features_level_path = '//input[@value="MAIN_FEATURES_LEVEL"]/following::input[@type="text"]';
		$main_features_level = $this->byXPath($main_features_level_path);
		$main_features_level->clear();
		$main_features_level->value('2');
		$this->byName('update')->click();
		// Page reloaded, we need a new XPath
		$main_features_level = $this->byXPath($main_features_level_path);
		return $this->assertEquals('2', $main_features_level->value(), "MAIN_FEATURES_LEVEL value is 2");
	}

	/**
	 * Test enabling the module
	 *
	 * @depends testEnableDeveloperMode
	 * @return bool
	 */
	public function testModuleEnabled()
	{
		$this->url('/admin/modules.php');
		$this->authenticate();
		$module_status_image_path = '//a[contains(@href, "'.self::$module_id.'")]/img';
		$module_status_image = $this->byXPath($module_status_image_path);
		if (strstr($module_status_image->attribute('src'), 'switch_off.png')) {
			// Enable the module
			$this->byHref('modMultiSalaryInput')->click();
		} else {
			// Disable the module
			$this->byHref('modMultiSalaryInput')->click();
			// Reenable the module
			$this->byHref('modMultiSalaryInput')->click();
		}
		// Page reloaded, we need a new Xpath
		$module_status_image = $this->byXPath($module_status_image_path);
		return $this->assertContains('switch_on.png', $module_status_image->attribute('src'), "Module enabled");
	}

	/**
	 * Test access to the configuration page
	 *
	 * @depends testModuleEnabled
	 * @return bool
	 */
	public function testConfigurationPage()
	{
		$this->url('/custom/multisalaryinput/admin/setup.php');
		$this->authenticate();
		return $this->assertContains('multisalaryinput/admin/setup.php', $this->url(), 'Configuration page');
	}

	/**
	 * Test access to the about page
	 *
	 * @depends testConfigurationPage
	 * @return bool
	 */
	public function testAboutPage()
	{
		$this->url('/custom/multisalaryinput/admin/about.php');
		$this->authenticate();
		return $this->assertContains('multisalaryinput/admin/about.php', $this->url(), 'About page');
	}

	/**
	 * Test about page is rendering Markdown
	 *
	 * @depends testAboutPage
	 * @return bool
	 */
	public function testAboutPageRendersMarkdownReadme()
	{
		$this->url('/custom/multisalaryinput/admin/about.php');
		$this->authenticate();
		return $this->assertEquals(
			'Dolibarr Module Template (aka My Module)',
			$this->byTag('h1')->text(),
			"Readme title"
		);
	}

	/**
	 * Test box is properly declared
	 *
	 * @depends testModuleEnabled
	 * @return bool
	 */
	public function testBoxDeclared()
	{
		$this->url('/admin/boxes.php');
		$this->authenticate();
		return $this->assertContains('multisalaryinputwidget1', $this->source(), "Box enabled");
	}

	/**
	 * Test trigger is properly enabled
	 *
	 * @depends testModuleEnabled
	 * @return bool
	 */
	public function testTriggerDeclared()
	{
		$this->url('/admin/triggers.php');
		$this->authenticate();
		return $this->assertContains(
			'interface_99_modMultiSalaryInput_MultiSalaryInputTriggers.class.php',
			$this->byTag('body')->text(),
			"Trigger declared"
		);
	}

	/**
	 * Test trigger is properly declared
	 *
	 * @depends testTriggerDeclared
	 * @return bool
	 */
	public function testTriggerEnabled()
	{
		$this->url('/admin/triggers.php');
		$this->authenticate();
		return $this->assertContains(
			'tick.png',
			$this->byXPath('//td[text()="interface_99_modMultiSalaryInput_MyTrigger.class.php"]/following::img')->attribute('src'),
			"Trigger enabled"
		);
	}

	/**
	 * Verify post conditions
	 * @return void
	 */
	protected function assertPostConditions()
	{
	}

	/**
	 * Unit test teardown
	 * @return void
	 */
	public function tearDown()
	{
	}

	/**
	 * Global test teardown
	 * @return void
	 */
	public static function tearDownAfterClass()
	{
	}
}


multisalaryinput.php
<?php

require 'dolibarr.inc.php';
require_once DOL_DOCUMENT_ROOT . '/core/lib/date.lib.php';
require_once DOL_DOCUMENT_ROOT . '/salaries/class/salary.class.php';
require_once DOL_DOCUMENT_ROOT . '/salaries/class/paymentsalary.class.php';
require_once DOL_DOCUMENT_ROOT . '/user/class/user.class.php';
require_once DOL_DOCUMENT_ROOT . '/user/class/usergroup.class.php';
require_once DOL_DOCUMENT_ROOT . '/core/class/html.formprojet.class.php';
require_once MULTISALARY_DOCUMENT_ROOT . '/lib/multisalaryinput.lib.php';

// Load translation files required by the page
$langs->loadLangs(array("multisalaryinput@multisalaryinput", "bills", "users", "salaries"));

if (!empty($conf->projet->enabled)) {
    $langs->load("projects");
}

$id = GETPOSTINT('id');
$ref = GETPOST('ref', 'alpha');
$action = GETPOST('action', 'aZ09');
$cancel = GETPOST('cancel', 'aZ09');
$note = GETPOST('note', 'restricthtml');
$label = GETPOST('label', 'alphanohtml');

$employees = GETPOST("employees", "array");
$usergroup = GETPOST("usergroup", "int");

$accountid = GETPOSTISSET('accountid') ? GETPOST('accountid', 'int') : 0;
$paymenttype = GETPOSTISSET('paymenttype') ? GETPOST('paymenttype', 'int') : 0;
$numpayment = GETPOSTISSET('num_payment') ? GETPOST('num_payment', 'int') : 0;
$projectid = GETPOSTISSET('fk_project') ? GETPOST('fk_project', 'int') : 0;

if (GETPOSTISSET('auto_create_paiement') || $action === 'add-multiple' || $action === 'save-multiple') {
    $auto_create_paiement = GETPOST("auto_create_paiement", "int");
} else {
    $auto_create_paiement = empty($conf->global->CREATE_NEW_SALARY_WITHOUT_AUTO_PAYMENT);
}

if (GETPOSTISSET('closepaidsalary') || $action === 'add-multiple' || $action === 'save-multiple') {
    $closepaidsalary = GETPOST("closepaidsalary", "int");
} else {
    $closepaidsalary = 1;
}

$datep = dol_mktime(12, 0, 0, GETPOST("datepmonth", 'int'), GETPOST("datepday", 'int'), GETPOST("datepyear", 'int'));
$datev = dol_mktime(12, 0, 0, GETPOST("datevmonth", 'int'), GETPOST("datevday", 'int'), GETPOST("datevyear", 'int'));
$datesp = dol_mktime(0, 0, 0, GETPOST("datespmonth", 'int'), GETPOST("datespday", 'int'), GETPOST("datespyear", 'int'));
$dateep = dol_mktime(23, 59, 59, GETPOST("dateepmonth", 'int'), GETPOST("dateepday", 'int'), GETPOST("dateepyear", 'int'));

$employeesSalaryAmount = GETPOST('employees_salary_amount', 'array');

$object = new Salary($db);
$extrafields = new ExtraFields($db);

$childids = $user->getAllChildIds(1);

// fetch optionals attributes and labels
$extrafields->fetch_name_optionals_label($object->table_element);

// Initialize technical object to manage hooks of page. Note that conf->hooks_modules contains array of hook context
$hookmanager->initHooks(array('multisalarycard', 'salarycard', 'globalcard'));

restrictedArea($user, 'salaries', $object->id, 'salary', '');

$title = $langs->trans('MultiSalaryInput') . " - " . $langs->trans('Card');
$help_url = "";

$formproject = new FormProjets($db);

/**
 * Actions
 */
// Note that $action and $object may have been modified by some hooks
$reshook = $hookmanager->executeHooks('doActions', [], $object, $action);

if ($reshook < 0) {
    setEventMessages($hookmanager->error, $hookmanager->errors, 'errors');
}

if (empty($reshook)) {
    $error = 0;

    switch ($action) {
        case 'add-multiple':
            $error = 0;
            if (empty($employees) && $usergroup <= 0) {
                setEventMessages($langs->trans("ErrorEmployeesOrUserGroupRequired"), null, 'errors');
                $error++;
            }

            if (!empty($auto_create_paiement) && empty($datep)) {
                $datePaymentMsg = $langs->trans("ErrorFieldRequired", $langs->transnoentitiesnoconv("DATE_PAIEMENT"));
                setEventMessages($datePaymentMsg, null, 'errors');
                $error++;
            }

            if (empty($datesp) || empty($dateep)) {
                $dateMsg = $langs->trans("ErrorFieldRequired", $langs->transnoentitiesnoconv("Date"));
                setEventMessages($dateMsg, null, 'errors');
                $error++;
            }

            if (!empty($auto_create_paiement) && (empty($paymenttype) || $paymenttype < 0)) {
                setEventMessages(
                        $langs->trans("ErrorFieldRequired", $langs->transnoentitiesnoconv("PaymentMode")), null,
                        'errors');
                $error++;
            }

            if (!empty($conf->banque->enabled) && !empty($auto_create_paiement) && !$accountid > 0) {
                $bankAccountMsg = $langs->trans("ErrorFieldRequired", $langs->transnoentitiesnoconv("BankAccount"));
                setEventMessages($bankAccountMsg, null, 'errors');
                $error++;
            }

            if ($error > 0) {
                $action = '';
                break;
            }

            break;

        case 'save-multiple':
            $error = 0;

            if (empty($employeesSalaryAmount)) {
                setEventMessages($langs->trans("ErrorMissingSalary"), null, 'errors');
                $action = 'add-multiple';
                break;
            }

            $db->begin();

            foreach ($employeesSalaryAmount as $employeeId => $employeeSalaryAmount) {
                if (empty($employeeSalaryAmount)) {
                    setEventMessages($langs->trans("ErrorMissingSalary"), null, 'errors');
                    $error++;
                    break;
                }

                if (floatval($employeeSalaryAmount) <= 0) {
                    setEventMessages($langs->trans("ErrorSalaryShouldBeAPositiveNumber"), null, 'errors');
                    $error++;
                    break;
                }

                $salary = new Salary($db);

                $salary->amount = floatval($employeeSalaryAmount);
                $salary->accountid = $accountid;
                $salary->fk_user = $employeeId;
                $salary->label = $label;
                $salary->datev = $datev;
                $salary->datep = $datep;
                $salary->datesp = $datesp;
                $salary->dateep = $dateep;
                $salary->note = $note;
                $salary->type_payment = ($paymenttype > 0 ? $paymenttype : 0);
                $salary->fk_user_author = $user->id;
                $salary->fk_project = $projectid;

                // Set user current salary as ref salary for the payment
                $fuser = new User($db);
                $fuser->fetch(GETPOST("fk_user", "int"));
                $salary->salary = $fuser->salary;

                // Fill array 'array_options' with data from add form
                $retSetOptionals = $extrafields->setOptionalsFromPost(null, $salary);

                if ($retSetOptionals < 0) {
                    $error++;
                    setEventMessages($extrafields->error, $extrafields->errors, 'errors');
                    break;
                }

                $retCreate = $salary->create($user);

                if ($retCreate < 0) {
                    setEventMessages($salary->error, $salary->errors, 'errors');
                    $error++;
                    break;
                }

                if (empty($auto_create_paiement)) {
                    continue;
                }

                // Create a line of payments
                $paiement = new PaymentSalary($db);
                $paiement->chid = $salary->id;
                $paiement->datepaye = $datep;
                $paiement->datev = $datev;
                $paiement->amounts = array($salary->id => floatval($employeeSalaryAmount)); // Tableau de montant
                $paiement->paiementtype = $paymenttype;
                $paiement->num_payment = $numpayment;
                $paiement->note = $note;

                $paymentid = $paiement->create($user, $closepaidsalary);

                if ($paymentid < 0) {
                    $error++;
                    setEventMessages($paiement->error, $paiement->errors, 'errors');
                    break;
                }

                $result = $paiement->addPaymentToBank($user, 'payment_salary', '(SalaryPayment)', $accountid, '', '');

                if ($result < 0) {
                    $error++;
                    setEventMessages($paiement->error, $paiement->errors, 'errors');
                    break;
                }
            }

            if (empty($error)) {
                $db->commit();
                header("Location: " . dol_buildpath('/salaries/list.php', 1));
                exit;
            }

            $db->rollback();
            $action = 'add-multiple';
            break;

        default:
            break;
    }
}



/**
 * Data
 */
$allEmployeesArray = [];
$errors = [];
$ret = getEmployeeArray($allEmployeesArray, $errors);

if ($ret < 0) {
    llxHeader("", $title, $help_url);
    dol_print_error($db, '', $errors);
    llxFooter();
}

if ($action == 'add-multiple') {
    $employeesArray = [];

    if (!empty($employees)) {
        foreach ($employees as $employeeId) {
            $employee = new User($db);

            $retFetch = $employee->fetch($employeeId);

            if ($retFetch < 0) {
                llxHeader("", $title, $help_url);
                dol_print_error($db, $employee->error, $employee->errors);
                llxFooter();
                exit;
            } elseif (empty($retFetch)) {
                setEventMessage('EmployeeNotFound', 'error');
                $action = '';
            }

            $employeesArray[] = $employee;
        }
    }

    if ($usergroup > 0) {
        $userGroupObject = new UserGroup($db);

        $userGroupObject->id = $usergroup;
        $userList = $userGroupObject->listUsersForGroup('COALESCE(employee, 0) <> 0');

        if (is_numeric($userList) && $userList < 0) {
            llxHeader("", $title, $help_url);
            dol_print_error($db, $userGroupObject->error, $userGroupObject->errors);
            llxFooter();
            exit;
        }

        if (empty($employeesArray)) {
            $employeesArray = $userList;
        } else {
            foreach ($userList as $groupuser) {
                $userFound = false;

                foreach ($employeesArray as $selectedEmployee) {
                    if ($groupuser->id == $selectedEmployee->id) {
                        $userFound = true;
                        break;
                    }
                }

                if (!$userFound) {
                    $employeesArray [] = $groupuser;
                }
            }
        }
    }
}


/*
 * 	View
 */

$form = new Form($db);

$year_current = strftime("%Y", dol_now());
$pastmonth = strftime("%m", dol_now()) - 1;
$pastmonthyear = $year_current;
if ($pastmonth == 0) {
    $pastmonth = 12;
    $pastmonthyear--;
}

if (empty($datesp) || empty($dateep)) { // We define  default date_start and date_end
    $datesp = dol_get_first_day($pastmonthyear, $pastmonth, false); 
    $dateep = dol_get_last_day($pastmonthyear, $pastmonth, false);
}

llxHeader("", $title, $help_url);

print load_fiche_titre($langs->trans("NewSalaries"), '', 'salary');

print dol_get_fiche_head('', '');

if ($action != 'add-multiple') {
    require MULTISALARY_DOCUMENT_ROOT . '/tpl/multisalaryinput_firstpage.tpl.php';
} else {
    require MULTISALARY_DOCUMENT_ROOT . '/tpl/multisalaryinput_secondpage.tpl.php';
}

print dol_get_fiche_end();

llxFooter();
exit;


lib/multisalaryinput.lib.php
<?php
/* Copyright (C) 2023 SuperAdmin
 *
 * This program is free software: you can redistribute it and/or modify
 * it under the terms of the GNU General Public License as published by
 * the Free Software Foundation, either version 3 of the License, or
 * (at your option) any later version.
 *
 * This program is distributed in the hope that it will be useful,
 * but WITHOUT ANY WARRANTY; without even the implied warranty of
 * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
 * GNU General Public License for more details.
 *
 * You should have received a copy of the GNU General Public License
 * along with this program.  If not, see <https://www.gnu.org/licenses/>.
 */

/**
 * \file    multisalaryinput/lib/multisalaryinput.lib.php
 * \ingroup multisalaryinput
 * \brief   Library files with common functions for MultiSalaryInput
 */

/**
 * Prepare admin pages header
 *
 * @return array
 */
function multisalaryinputAdminPrepareHead()
{
	global $langs, $conf;

	$langs->load("multisalaryinput@multisalaryinput");

	$h = 0;
	$head = array();

	$head[$h][0] = dol_buildpath("/multisalaryinput/admin/setup.php", 1);
	$head[$h][1] = $langs->trans("Settings");
	$head[$h][2] = 'settings';
	$h++;

	/*
	$head[$h][0] = dol_buildpath("/multisalaryinput/admin/myobject_extrafields.php", 1);
	$head[$h][1] = $langs->trans("ExtraFields");
	$head[$h][2] = 'myobject_extrafields';
	$h++;
	*/

	$head[$h][0] = dol_buildpath("/multisalaryinput/admin/about.php", 1);
	$head[$h][1] = $langs->trans("About");
	$head[$h][2] = 'about';
	$h++;

	// Show more tabs from modules
	// Entries must be declared in modules descriptor with line
	//$this->tabs = array(
	//	'entity:+tabname:Title:@multisalaryinput:/multisalaryinput/mypage.php?id=__ID__'
	//); // to add new tab
	//$this->tabs = array(
	//	'entity:-tabname:Title:@multisalaryinput:/multisalaryinput/mypage.php?id=__ID__'
	//); // to remove a tab
	complete_head_from_modules($conf, $langs, null, $head, $h, 'multisalaryinput@multisalaryinput');

	complete_head_from_modules($conf, $langs, null, $head, $h, 'multisalaryinput@multisalaryinput', 'remove');

	return $head;
}


function getEmployeeArray(&$employeesArray, &$errors)
{
    global $db, $user, $conf, $hookmanager;
    // Forge request to select users
    $sql = "SELECT DISTINCT u.rowid, u.lastname as lastname, u.firstname";

    if (!empty($conf->multicompany->enabled) && $conf->entity == 1 && $user->admin && !$user->entity) {
        $sql .= ", e.label";
    }

    $sql .= " FROM " . MAIN_DB_PREFIX . "user as u";
    
    if (!empty($conf->multicompany->enabled) && $conf->entity == 1 && $user->admin && !$user->entity) {
        $sql .= " LEFT JOIN " . MAIN_DB_PREFIX . "entity as e ON e.rowid = u.entity";
        $sql .= " WHERE u.entity IS NOT NULL";
    } else {
        if (!empty($conf->multicompany->enabled) && !empty($conf->global->MULTICOMPANY_TRANSVERSE_MODE)) {
            $sql .= " LEFT JOIN " . MAIN_DB_PREFIX . "usergroup_user as ug";
            $sql .= " ON ug.fk_user = u.rowid";
            $sql .= " WHERE ug.entity = " . $conf->entity;
        } else {
            $sql .= " WHERE u.entity IN (0, " . $conf->entity . ")";
        }
    }

    $sql .= " AND COALESCE(u.employee,0) <> 0";

    if (!empty($conf->global->USER_HIDE_INACTIVE_IN_COMBOBOX)) {
        $sql .= " AND COALESCE(u.statut,0) <> 0";
    }

    //Add hook to filter on user (for exemple on usergroup define in custom modules)
    $reshook = $hookmanager->executeHooks('addSQLWhereFilterOnSelectUsers', array());
    
    if (!empty($reshook)) {
        $sql .= $hookmanager->resPrint;
    }

    // MAIN_FIRSTNAME_NAME_POSITION is 0 means firstname+lastname
    if (empty($conf->global->MAIN_FIRSTNAME_NAME_POSITION)) {
        $sql .= " ORDER BY u.statut DESC, u.firstname ASC, u.lastname ASC";
    } else {
        $sql .= " ORDER BY u.statut DESC, u.lastname ASC, u.firstname ASC";
    }

    $resql = $db->query($sql);
    
    if (!$resql) {
        $errors = $db->lasterror();
        return -1;
    }


    $num = $db->num_rows($resql);

    for ($i = 0; $i < $num; $i++) {
        $obj = $db->fetch_object($resql);
        $employeesArray[$obj->rowid] = $obj->firstname . ' ' . $obj->lastname;
    }

    return 1;
}


tpl/multisalaryinput_secondpage.tpl.php
<form name="multi-salary-form" action="<?= $_SERVER["PHP_SELF"] ?>" method="post">
    <input type="hidden" name="action" value="save-multiple"/>
    <input type="hidden" name="token" value="<?= newToken() ?>"/>
    <div style="display:none">
        <input type="hidden" name="label"  id="label" value="<?= $label ? : $langs->trans("Salary") ?>">
        <?= $formproject->select_projects(-1, $projectid, 'fk_project', 0, 0, 1, 1, 0, 0, 0, '', 1) ?>
        <?= $form->select_comptes($accountid, "accountid", 0, '', 1) ?>
        <?= $form->select_types_paiements($paymenttype, "paymenttype", '', 0, 1, 0, 0, 1, '', 1) ?>*
        <?= $form->selectDate($datesp, "datesp", '', '', '', 'add') ?>
        <?= $form->selectDate($dateep, "dateep", '', '', '', 'add') ?>
        <?= $form->selectDate((empty($datep) ? '' : $datep), "datep", 0, 0, 0, 'add', 1, 1) ?>
        <?= $form->selectDate((empty($datev) ? -1 : $datev), "datev", '', '', '', 'add', 1, 1) ?>
        <input id="num_payment" name="num_payment" type="text" value="<?= $numpayment ?>">
        <input id="auto_create_paiement" name="auto_create_paiement" value="<?= $auto_create_paiement ?>">
        <input id="closepaidsalary" name="closepaidsalary" value="<?= $closepaidsalary ?>">
        <textarea name="note" wrap="soft" cols="60" rows="<?= ROWS_3 ?>"><?= $note ?></textarea>
        <?= $form->multiselectarray('employees', $allEmployeesArray, $employees, '', 0, '', 0, '50%')?>
        <?= $form->select_dolgroups($usergroup, 'usergroup', 1) ?>
<?php $reshook = $hookmanager->executeHooks('formObjectOptions', $parameters, $object, $action); ?>
        <?=$hookmanager->resPrint; ?>
<?php if (empty($reshook)) : ?>
        <?= $object->showOptionals($extrafields, 'create'); ?>
<?php endif; ?>
        
    </div>
    <table class="border centpercent">
<?php foreach ($employeesArray as $employee): ?>
        <tr>
            <td><?= $employee->getNomUrl() ?></td>
            <td class="left">
                <input name="employees_salary_amount[<?= $employee->id ?>]" 
                       value="<?= $employeesSalaryAmount[$employee->id] ? : '' ?>"
                       placeholder="€" />
            </td>
        </tr>
<?php endforeach; ?>
    </table>
    <br>
    <br>        
    <div class="center">
        <input type="submit" class="button button-save" name="save" value="<?= $langs->trans('SaveSalaries') ?>">
        <a class="butAction" href="<?= $_SERVER["PHP_SELF"] ?>"><?= $langs->trans('Cancel') ?></a>
    </div>
</form>

tpl/multisalaryinput_firstpage.tpl.php
<form name="multi-salary-form" action="<?= $_SERVER["PHP_SELF"] ?>" method="post">
    <input type="hidden" name="action" value="add-multiple"/>
    <input type="hidden" name="token" value="<?= newToken() ?>"/>
    <table class="border centpercent">
        <tr>
            <td>
                <?= $form->editfieldkey('Employees', 'employees', '', $object, 0, 'string', '', 1) ?>
            </td>
            <td>
                <?= $form->multiselectarray('employees', $allEmployeesArray, $employees, '', 0, '', 0, '50%')
                    . $form->select_dolgroups($usergroup, 'usergroup', 1) ?>
            </td>
        </tr>
        <tr>
            <td>
                <?= $form->editfieldkey('Label', 'label', '', $object, 0, 'string', '', 1) ?>
            </td>
            <td>
                <input 
                    name="label" 
                    id="label" 
                    class="minwidth300" 
                    value="<?= ($label ? : $langs->trans("Salary")) ?>">
            </td>
        </tr>
        <tr>
            <td>
                <?= $form->editfieldkey('DateStartPeriod', 'datesp', '', $object, 0, 'string', '', 1) ?>
            </td>
            <td>
                <?= $form->selectDate($datesp, "datesp", '', '', '', 'add') ?>
            </td>
        </tr>
        <tr>
            <td>
                <?= $form->editfieldkey('DateEndPeriod', 'dateep', '', $object, 0, 'string', '', 1) ?>
            </td>
            <td>
                <?= $form->selectDate($dateep, "dateep", '', '', '', 'add') ?>
            </td>
        </tr>
<?php if (!empty($conf->projet->enabled)) : ?>
        <tr>
            <td>
                <?= $langs->trans("Project") ?>
            </td>
            <td>
                <?= img_picto('', 'project', 'class="pictofixedwidth"')
            . $formproject->select_projects(-1, $projectid, 'fk_project', 0, 0, 1, 1, 0, 0, 0, '', 1) ?>
            </td>
        </tr>
<?php endif; ?>
        <tr>
            <td class="tdtop">
                <?= $langs->trans("Comments") ?>
            </td>
            <td class="tdtop">
                <textarea name="note" wrap="soft" cols="60" rows="<?= ROWS_3 ?>"><?= $note ?></textarea>
            </td>
        </tr>
        <tr>
            <td colspan="2"><hr></td>
        </tr>
        <tr>
            <td>
                <label for="auto_create_paiement"><?= $langs->trans('AutomaticCreationPayment') ?></label>
            </td>
            <td>
                <input 
                    id="auto_create_paiement" 
                    name="auto_create_paiement" 
                    type="checkbox" 
                    <?= (empty($auto_create_paiement) ? '' : 'checked="checked"') ?> 
                    value="1">
            </td>
        </tr>
<?php if (!empty($conf->banque->enabled)) : ?>
        <tr>
            <td id="label_fk_account">
                <?= $form->editfieldkey('BankAccount', 'selectaccountid', '', $object, 0, 'string', '', 1) ?>
            </td>
            <td> 
                <?= img_picto('', 'bank_account', 'class="paddingrighonly"')
            . $form->select_comptes($accountid, "accountid", 0, '', 1) ?>
            </td>
        </tr>
<?php endif; ?>
        <tr>
            <td id="label_type_payment">
                <?= $form->editfieldkey('PaymentMode', 'selectpaymenttype', '', $object, 0, 'string', '', 1) ?>
            </td>
            <td>
                <?= $form->select_types_paiements($paymenttype, "paymenttype", '', 0, 1, 0, 0, 1, '', 1) ?>
            </td>
        </tr>
        <tr class="hide_if_no_auto_create_payment">
            <td>
                <?= $form->editfieldkey('DatePayment', 'datep', '', $object, 0, 'string', '', 1) ?>
            </td>
            <td>
                <?= $form->selectDate((empty($datep) ? '' : $datep), "datep", 0, 0, 0, 'add', 1, 1) ?>
            </td>
        </tr>
        <tr class="hide_if_no_auto_create_payment">
            <td>
                <?= $form->editfieldkey('DateValue', 'datev', '', $object, 0) ?>
            </td>
            <td>
                <?= $form->selectDate((empty($datev) ? -1 : $datev), "datev", '', '', '', 'add', 1, 1) ?>
            </td>
        </tr>

<?php if (!empty($conf->banque->enabled)) : ?>
        <tr class="hide_if_no_auto_create_payment">
            <td>
                <label for="num_payment"><?= $langs->trans('Numero') ?> <em>(<?= $langs->trans("ChequeOrTransferNumber") ?>)</em></label>
            </td>
            <td>
                <input name="num_payment" id="num_payment" type="text" value="<?= $numpayment ?>">
            </td>
        </tr>
<?php endif; ?>
        
<?php $reshook = $hookmanager->executeHooks('formObjectOptions', $parameters, $object, $action); ?>
        <?=$hookmanager->resPrint; ?>
<?php if (empty($reshook)) : ?>
        <?= $object->showOptionals($extrafields, 'create'); ?>
<?php endif; ?>

    </table>
    <div class="center">
        <div class="hide_if_no_auto_create_payment paddingbottom">
            <input type="checkbox" <?= ($closepaidsalary > 0 ? 'checked' : '') ?> value="<?= $closepaidsalary ?>" name="closepaidsalary"><?= $langs->trans("ClosePaidSalaryAutomatically") ?>
        </div>
    </div>
    <div class="center">
        <input class="butAction" type="submit" value="<?= $langs->trans("Next") ?>"/>
    </div>
</form>